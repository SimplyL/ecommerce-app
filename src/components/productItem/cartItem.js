import React from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import { PRODUCT_TYPE } from '../../constants/types';
import { removeProduct } from '../../modules/shoppingCart';
import {
  CartItemContainer,
  Image,
  ImageContainer,
  Title,
  Label,
  PairContainer,
  PairValue,
  RemoveButton,
  CartItemContent,
  CartItemDetails,
} from './productItem.styles';

const calculateFullPrice = (price, quantity) => {
  const fullPrice = quantity * price;
  return parseFloat(fullPrice).toFixed(2);
}

const renderLabelValuePair = (label, value, hasSymbol) => (
  <PairContainer>
    <Label>
      {label}
    </Label>
    <PairValue>
      {value}
      {hasSymbol && <span> &euro;</span>}
    </PairValue>
  </PairContainer>
);

export const CartItem = ({ product, quantity, removeProduct }) => {
  const { id, name, price, image } = product;

  return (
    <CartItemContainer isLeft>
      <ImageContainer>
        <Image src={require(`images/${image}`)} alt="img" />
      </ImageContainer>
      <CartItemContent>
        <Title>{name}</Title>
        <CartItemDetails>
          {renderLabelValuePair('Item price:', price, true)}
          {renderLabelValuePair('Quantity:', quantity)}
          {renderLabelValuePair('Full price:', calculateFullPrice(price, quantity), true)}
          <RemoveButton onClick={() => removeProduct(id)}>Remove item</RemoveButton>
        </CartItemDetails>
      </CartItemContent>
    </CartItemContainer>
  );
};

CartItem.propTypes = {
  product: PRODUCT_TYPE.isRequired,
  removeProduct: PropTypes.func.isRequired,
};

export default connect(null, { removeProduct })(CartItem);