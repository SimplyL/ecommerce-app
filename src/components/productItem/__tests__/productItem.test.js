import React from 'react';
import { shallow } from 'enzyme';

import { CartItem } from '../cartItem';
import ProductDescription from '../productDescription';
import ProductItem from '../index';
import { product } from './__mocks__/product';
import PATHS from '../../../constants/paths';

describe('CartItem component', () => {
  let wrapper;
  const mockFn = jest.fn();

  beforeEach(() => {
    wrapper = shallow(<CartItem product={product} removeProduct={mockFn} />);
  });

  it('should render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should call remove function with correct parameters', () => {
    wrapper.find('RemoveButton').simulate('click');
    expect(mockFn).toBeCalledWith("0");
  });
});

describe('ProductDescription component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ProductDescription product={product} />);
  });

  it('should render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });
});

describe('ProductItem component', () => {
  let wrapper;

  beforeEach(() => {
    wrapper = shallow(<ProductItem product={product} />);
  });

  it('should render without crashing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render hyperlink with correct route', () => {
    expect(wrapper.find('HyperLink').props().to).toEqual(`${PATHS.product}/0`);
  });
});
