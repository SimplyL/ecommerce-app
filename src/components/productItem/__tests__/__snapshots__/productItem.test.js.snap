// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CartItem component should render without crashing 1`] = `
<CartItemContainer
  isLeft={true}
>
  <ImageContainer>
    <Image
      alt="img"
      src="test-file-stub"
    />
  </ImageContainer>
  <CartItemContent>
    <Title>
      Model
    </Title>
    <CartItemDetails>
      <PairContainer>
        <Label>
          Item price:
        </Label>
        <styled.span>
          4.99
          <span>
             €
          </span>
        </styled.span>
      </PairContainer>
      <PairContainer>
        <Label>
          Quantity:
        </Label>
        <styled.span />
      </PairContainer>
      <PairContainer>
        <Label>
          Full price:
        </Label>
        <styled.span>
          NaN
          <span>
             €
          </span>
        </styled.span>
      </PairContainer>
      <RemoveButton
        onClick={[Function]}
      >
        Remove item
      </RemoveButton>
    </CartItemDetails>
  </CartItemContent>
</CartItemContainer>
`;

exports[`ProductDescription component should render without crashing 1`] = `
<DescriptionContainer>
  <ImageContainer>
    <Image
      alt="img"
      src="test-file-stub"
    />
  </ImageContainer>
  <styled.div>
    <Title>
      Model
    </Title>
    <InputContainer>
      <PriceLabel>
        4.99
         €
      </PriceLabel>
      <AddToCartInput
        id="0"
      />
    </InputContainer>
    <Description>
      Lorem ipsum dolor sit amet, consectetur adipiscing elit.
    </Description>
  </styled.div>
</DescriptionContainer>
`;

exports[`ProductItem component should render without crashing 1`] = `
<Container>
  <ImageContainer>
    <Image
      alt="img"
      src="test-file-stub"
    />
  </ImageContainer>
  <HyperLink
    to="/product/0"
  >
    Model
  </HyperLink>
  <PriceLabel>
    4.99
     €
  </PriceLabel>
  <ConnectFunction
    id="0"
  />
</Container>
`;
